File: win-bot/src/botworker/stomp-client.js
'use strict';

const WebSocket = require('ws');

const NULL_BYTE = '\x00';

function buildFrame(command, headers = {}, body = '') {
  let frame = command + '\n';
  for (const [k, v] of Object.entries(headers)) frame += `${k}:${v}\n`;
  frame += '\n' + body + NULL_BYTE;
  return frame;
}

function parseFrame(raw) {
  const nullIdx = raw.indexOf('\x00');
  const text = nullIdx >= 0 ? raw.slice(0, nullIdx) : raw;
  // Normalize line endings: strip \r so \r\n works the same as \n
  const lines = text.split('\n').map(l => l.replace(/\r$/, ''));
  if (!lines.length || !lines[0].trim()) return null;

  const command = lines[0].trim();
  const headers = {};
  let i = 1;
  while (i < lines.length && lines[i].trim() !== '') {
    const colon = lines[i].indexOf(':');
    if (colon > 0) headers[lines[i].slice(0, colon).trim()] = lines[i].slice(colon + 1).trim();
    i++;
  }
  const body = lines.slice(i + 1).join('\n');
  return { command, headers, body };
}

class StompClient {
  constructor(serverUrl, authClient) {
    this.serverUrl = serverUrl;
    this.authClient = authClient;
    this.ws = null;
    this.connected = false;
    this.subscriptions = new Map(); // destination -> callback
    this.onConnectedCallbacks = [];
    this.onAssignCallback = null;
    this.accessToken = null;
    this.refreshToken = null;
    this._stopped = false;
    this._subCounter = 0;
    this._subIds = new Map(); // destination -> subId
  }

  onConnected(cb)  { this.onConnectedCallbacks.push(cb); }
  onAssign(cb)     { this.onAssignCallback = cb; }

  start() { this._loop(); }

  async _loop() {
    let attempt = 0;
    while (!this._stopped) {
      try {
        const tokens = await this.authClient.loginOrRegister();
        this.accessToken  = tokens.accessToken;
        this.refreshToken = tokens.refreshToken;
        attempt = 0;
        await this._connect();
      } catch (e) {
        console.error('[STOMP] Connection error:', e.message);
      }
      if (this._stopped) break;
      const delay = Math.min(Math.pow(2, attempt) * 1000, 60_000);
      console.log(`[STOMP] Reconnecting in ${delay / 1000}s...`);
      attempt++;
      await new Promise(r => setTimeout(r, delay));
    }
  }

  _connect() {
    return new Promise((resolve) => {
      const ws = new WebSocket(this.serverUrl);
      this.ws = ws;

      ws.on('open', () => {
        ws.send(buildFrame('CONNECT', {
          'accept-version': '1.1,1.2',
          'heart-beat': '10000,10000',
          'Authorization': `Bearer ${this.accessToken}`,
        }));
      });

      ws.on('message', (data) => {
        const raw = data.toString();
        if (raw === '\n') { ws.send('\n'); return; } // heartbeat pong
        const frame = parseFrame(raw);
        if (!frame) return;
        this._handleFrame(frame);
      });

      ws.on('close', (code, reason) => {
        console.warn(`[STOMP] Disconnected: ${code} ${reason}`);
        this.connected = false;
        resolve();
      });

      ws.on('error', (err) => {
        console.error('[STOMP] WS error:', err.message);
        this.connected = false;
        resolve();
      });
    });
  }

  _handleFrame(frame) {
    const { command, headers, body } = frame;

    if (command === 'CONNECTED') {
      this.connected = true;
      console.log('[STOMP] CONNECTED');

      // Subscribe to bot invite queue
      this._sendSubscribe('/user/queue/bot-invite', 'sub-bot-invite');

      // Re-subscribe all existing subscriptions
      for (const [dest, _] of this.subscriptions) {
        const sid = this._subIds.get(dest) || `sub-${++this._subCounter}`;
        this._subIds.set(dest, sid);
        this._sendSubscribe(dest, sid);
      }

      for (const cb of this.onConnectedCallbacks) {
        try { cb(); } catch (e) { console.error('[STOMP] onConnected error:', e); }
      }
      return;
    }

    if (command === 'MESSAGE') {
      const dest = headers.destination || '';
      let payload = {};
      if (body) { try { payload = JSON.parse(body); } catch {} }

      if (dest === '/user/queue/bot-invite' || dest.endsWith('/queue/bot-invite')) {
        if (this.onAssignCallback) this.onAssignCallback(payload);
        return;
      }

      for (const [topic, cb] of this.subscriptions) {
        if (dest === topic || dest.endsWith(topic)) { cb(payload); return; }
      }
      return;
    }

    if (command === 'ERROR') {
      console.error('[STOMP] ERROR frame:', headers.message, body);
    }
  }

  _sendSubscribe(dest, id) {
    if (this.ws && this.connected) {
      this.ws.send(buildFrame('SUBSCRIBE', { id, destination: dest }));
    }
  }

  subscribe(dest, cb) {
    if (!this._subIds.has(dest)) {
      const id = `sub-${++this._subCounter}`;
      this._subIds.set(dest, id);
      if (this.connected) this._sendSubscribe(dest, id);
    }
    this.subscriptions.set(dest, cb);
  }

  send(dest, payload) {
    if (!this.ws || !this.connected) {
      console.warn(`[STOMP] Cannot send to ${dest}: not connected`);
      return;
    }
    const body = JSON.stringify(payload);
    this.ws.send(buildFrame('SEND', {
      destination: dest,
      'content-type': 'application/json',
    }, body));
  }

  isConnected() { return this.connected; }
}

module.exports = { StompClient };

File: py-bot-worker/src/botworker/stomp_client.py
import json
import time
import logging
import threading
from typing import Callable, Dict, List, Optional
import websocket

logger = logging.getLogger("StompConnectionService")

class StompFrame:
    def __init__(self, command: str, headers: Dict[str, str], body: str = ""):
        self.command = command
        self.headers = headers
        self.body = body

    @classmethod
    def parse(cls, raw: str) -> Optional["StompFrame"]:
        if not raw or raw == "\n":
            return None
        lines = raw.rstrip("\x00").split("\n")
        if not lines:
            return None
        command = lines[0].strip()
        headers = {}
        idx = 1
        while idx < len(lines) and lines[idx]:
            header_line = lines[idx]
            if ":" in header_line:
                k, v = header_line.split(":", 1)
                headers[k.strip()] = v.strip()
            idx += 1
        body = "\n".join(lines[idx + 1:]) if idx + 1 < len(lines) else ""
        return cls(command, headers, body)

    def serialize(self) -> str:
        res = [self.command]
        for k, v in self.headers.items():
            res.append(f"{k}:{v}")
        res.append("")
        res.append(self.body)
        return "\n".join(res) + "\x00"

class StompConnectionService:
    def __init__(self, server_url: str, auth_client):
        self.server_url = server_url
        self.auth_client = auth_client
        self.ws: Optional[websocket.WebSocketApp] = None
        self.connected = False
        self.sub_id_counter = 0
        self.subscriptions: Dict[str, Tuple[str, Callable[[dict], None]]] = {} # topic -> (sub_id, callback)
        self.on_connected_callbacks: List[Callable[[], None]] = []
        self.on_assign_callback: Optional[Callable[[dict], None]] = None
        self.access_token: Optional[str] = None
        self.refresh_token: Optional[str] = None
        self.worker_thread: Optional[threading.Thread] = None
        self.stop_requested = False

    def add_on_connected_callback(self, cb: Callable[[], None]):
        self.on_connected_callbacks.append(cb)

    def set_on_assign_callback(self, cb: Callable[[dict], None]):
        self.on_assign_callback = cb

    def start(self):
        self.worker_thread = threading.Thread(target=self._run_loop, daemon=True)
        self.worker_thread.start()

    def _run_loop(self):
        attempt = 0
        while not self.stop_requested:
            try:
                self.auth_client.register()
                tokens = self.auth_client.login()
                self.access_token = tokens.access_token
                self.refresh_token = tokens.refresh_token
                attempt = 0

                headers = {}
                self.ws = websocket.WebSocketApp(
                    self.server_url,
                    header=headers,
                    on_open=self._on_ws_open,
                    on_message=self._on_ws_message,
                    on_error=self._on_ws_error,
                    on_close=self._on_ws_close
                )
                self.ws.run_forever(ping_interval=20, ping_timeout=10)
            except Exception as e:
                logger.error(f"STOMP connection loop error: {e}")

            attempt += 1
            delay = min(2 ** attempt, 60)
            logger.info(f"Reconnecting STOMP in {delay}s...")
            time.sleep(delay)

    def _on_ws_open(self, ws):
        logger.info(f"WebSocket opened to {self.server_url}, sending CONNECT frame...")
        stomp_headers = {
            "accept-version": "1.1,1.2",
            "heart-beat": "10000,10000"
        }
        if self.access_token:
            stomp_headers["Authorization"] = f"Bearer {self.access_token}"
        
        connect_frame = StompFrame("CONNECT", stomp_headers)
        ws.send(connect_frame.serialize())

    def _on_ws_message(self, ws, message):
        frame = StompFrame.parse(message)
        if frame is None:
            return

        if frame.command == "CONNECTED":
            self.connected = True
            logger.info("STOMP session CONNECTED!")
            
            # Subscribe to bot invite queue
            self._send_subscribe("/user/queue/bot-invite", "sub-bot-invite")

            # Resubscribe topics
            for topic, (sub_id, _) in list(self.subscriptions.items()):
                self._send_subscribe(topic, sub_id)

            for cb in self.on_connected_callbacks:
                try:
                    cb()
                except Exception as e:
                    logger.error(f"Error in on_connected callback: {e}")

        elif frame.command == "MESSAGE":
            dest = frame.headers.get("destination", "")
            body = frame.body
            payload = {}
            if body:
                try:
                    payload = json.loads(body)
                except Exception as e:
                    logger.warning(f"Failed to parse JSON body: {body}")

            if dest == "/user/queue/bot-invite" or dest.endswith("/queue/bot-invite"):
                if self.on_assign_callback:
                    self.on_assign_callback(payload)
            else:
                for topic, (_, cb) in list(self.subscriptions.items()):
                    if dest == topic or dest.endswith(topic):
                        cb(payload)

        elif frame.command == "ERROR":
            logger.error(f"STOMP ERROR frame received: {frame.headers.get('message')} | {frame.body}")

    def _on_ws_error(self, ws, error):
        logger.error(f"WebSocket transport error: {error}")
        self.connected = False

    def _on_ws_close(self, ws, close_status_code, close_msg):
        logger.warning(f"WebSocket closed: {close_status_code} - {close_msg}")
        self.connected = False

    def subscribe(self, topic: str, callback: Callable[[dict], None]):
        if topic not in self.subscriptions:
            self.sub_id_counter += 1
            sub_id = f"sub-{self.sub_id_counter}"
            self.subscriptions[topic] = (sub_id, callback)
            if self.connected:
                self._send_subscribe(topic, sub_id)
        else:
            sub_id, _ = self.subscriptions[topic]
            self.subscriptions[topic] = (sub_id, callback)

    def _send_subscribe(self, topic: str, sub_id: str):
        if self.ws and self.connected:
            frame = StompFrame("SUBSCRIBE", {
                "id": sub_id,
                "destination": topic
            })
            self.ws.send(frame.serialize())

    def send(self, destination: str, payload: dict):
        if self.ws and self.connected:
            body = json.dumps(payload)
            frame = StompFrame("SEND", {
                "destination": destination,
                "content-type": "application/json"
            }, body)
            self.ws.send(frame.serialize())
        else:
            logger.warning(f"Cannot send to {destination} - STOMP not connected")

    def is_connected(self) -> bool:
        return self.connected
